torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -240,6 +249,133 @@ class BatchDataBuffer {
   // the program to hang. This boolean is used to break this waiting condition.
   std::atomic<bool> stop_;
 };
+
+/// Select chunks for loading and define a sampling behavior.
+/// In a distributed setting, this selects a subset of the chunks depending on
+/// the provided num_replicas and rank parameters. The `next()` method of this
+/// class needs to be thread-safe as it will be called from different threads
+/// during chunk loading.
+class ChunkSelector {
+ public:
+  virtual ~ChunkSelector() = default;
+  ChunkSelector(size_t chunk_count, size_t num_replicas = 1, size_t rank = 0)
+      : chunk_count_(chunk_count),
+        num_replicas_(num_replicas),
+        rank_(rank),
+        epoch_(0) {
+    local_chunk_count_ = (size_t)std::ceil(chunk_count_ * 1.0 / num_replicas_);
+  }
+
+  /// Get the next chunk index for loading.
+  /// Note: this method needs to be thread-safe.
+  virtual optional<size_t> next() = 0;
+
+  /// Reset the chunk selector for a new enumeration.
+  virtual void reset() = 0;
+
+  /// Set the epoch for the current enumeration. This can be used to alter the
+  /// chunk selection and shuffling behavior.
+  void set_epoch(size_t epoch) {
+    epoch_ = epoch;
+  }
+
+  /// Return the number of chunks to be loaded. In the case of distributed
+  /// training, this is different to chunk_count as each loader needs to load
+  /// only a subset of chunks.
+  size_t local_chunk_count() {
+    return local_chunk_count_;
+  }
+
+ protected:
+  size_t epoch_;
+  size_t chunk_count_;
+  size_t num_replicas_;
+  size_t rank_;
+  size_t local_chunk_count_;
+};
+
+/// Select chunks randomly. The chunk order shuffled at each `reset()` call.
+class RandomChunkSelector : public ChunkSelector {
+ public:
+  RandomChunkSelector(
+      size_t chunk_count,
+      size_t num_replicas = 1,
+      size_t rank = 0)
+      : ChunkSelector(chunk_count, num_replicas, rank) {
+    size_t index_count =
+        num_replicas_ == 1 ? chunk_count_ : local_chunk_count_ * num_replicas_;
+    all_indices_.resize(index_count);
+    std::iota(std::begin(all_indices_), std::end(all_indices_), 0);
+    if (num_replicas_ > 1)
+    {
+      for (size_t i = chunk_count; i < index_count; ++i) {
+        all_indices_[i] =
+            i % chunk_count_; // we are adding some more chunks to make all
```

Comment:
Curious, so we will have duplicated chunk index and load duplicated chunk data ? Will this be a problem in training?

Reply:
Thank you for the great question. We figured an important solution from this.